hal/src/common/dmac/dma_controller.rs

```diff
@@ -0,0 +1,420 @@
+        // right now we blindly enable all priority levels
+        dmac.ctrl.modify(|_, w| {
+            w.lvlen3()
+                .set_bit()
+                .lvlen2()
+                .set_bit()
+                .lvlen1()
+                .set_bit()
+                .lvlen0()
+                .set_bit()
+        });
+
+        // Enable DMA controller
+        dmac.ctrl.modify(|_, w| w.dmaenable().set_bit());
+
+        Self { dmac }
+    }
+
+    /// Enable or disable priority level 0
+    #[inline(always)]
+    pub fn level_0_enabled(&mut self, enabled: bool) {
+        self.dmac.ctrl.modify(|_, w| w.lvlen0().bit(enabled));
+    }
+
+    /// Enable or disable priority level 1
+    #[inline(always)]
+    pub fn level_1_enabled(&mut self, enabled: bool) {
+        self.dmac.ctrl.modify(|_, w| w.lvlen1().bit(enabled));
+    }
+
+    /// Enable or disable priority level 2
+    #[inline(always)]
+    pub fn level_2_enabled(&mut self, enabled: bool) {
+        self.dmac.ctrl.modify(|_, w| w.lvlen2().bit(enabled));
+    }
+
+    /// Enable or disable priority level 3
+    #[inline(always)]
+    pub fn level_3_enabled(&mut self, enabled: bool) {
+        self.dmac.ctrl.modify(|_, w| w.lvlen3().bit(enabled));
+    }
+
+    /// Enable or disable Round-Robin Arbitration for priority level 0
+    #[inline(always)]
+    pub fn level_0_round_robin(&mut self, enabled: bool) {
+        self.dmac.prictrl0.modify(|_, w| w.rrlvlen0().bit(enabled));
+    }
+
+    /// Enable or disable Round-Robin Arbitration for priority level 1
+    #[inline(always)]
+    pub fn level_1_round_robin(&mut self, enabled: bool) {
+        self.dmac.prictrl0.modify(|_, w| w.rrlvlen1().bit(enabled));
+    }
+
+    /// Enable or disable Round-Robin Arbitration for priority level 2
+    #[inline(always)]
+    pub fn level_2_round_robin(&mut self, enabled: bool) {
+        self.dmac.prictrl0.modify(|_, w| w.rrlvlen2().bit(enabled));
+    }
+
+    /// Enable or disable Round-Robin Arbitration for priority level 3
+    #[inline(always)]
+    pub fn level_3_round_robin(&mut self, enabled: bool) {
+        self.dmac.prictrl0.modify(|_, w| w.rrlvlen3().bit(enabled));
+    }
+
+    /// Release the DMAC and return the register block
+    pub fn release(mut self, _pm: &mut PM) -> DMAC {
+        self.dmac.ctrl.modify(|_, w| w.dmaenable().clear_bit());
+
+        Self::swreset(&mut self.dmac);
+
+        #[cfg(any(feature = "samd11", feature = "samd21"))]
+        {
+            // Disable the DMAC clocking
+            _pm.apbbmask.modify(|_, w| w.dmac_().clear_bit());
+            _pm.ahbmask.modify(|_, w| w.dmac_().clear_bit());
+        }
+
+        // Release the DMAC
+        self.dmac
+    }
+
+    /// Issue a software reset to the DMAC and wait for reset to complete
+    #[inline(always)]
+    fn swreset(dmac: &mut DMAC) {
+        dmac.ctrl.modify(|_, w| w.swrst().set_bit());
+        while !dmac.ctrl.read().swrst().bit_is_clear() {}
```

Comment:
This is a nit, but I would prefer `bit_is_set` here. I tend to avoid `!` when possible. It's really narrow and hard to see. It's also a prefix operator, which then applies to the condition at the end of the line (`bit_is_clear`). I think using the `bit_is_set` function is much more readable.